### PR TITLE
Feature/tec 3385 recent past events

### PR DESCRIPTION
**Pull request by jesseeproductions**
_Ref:_ [TEC-3385](https://moderntribe.atlassian.net/browse/TEC-3385)

This is almost ready for an entire review, but has some todos to finish up. 

The recent past templates are not for review they are placeholders.

Besides those this can be reviewed to move it forward. 

Related PR:
https://github.com/moderntribe/tribe-common/pull/1332

### CODE REVIEW

File: src/Tribe/Views/V2/View.php [hunk 1]
```diff
@@ -1352,6 +1355,7 @@ public function has_next_event( array $events, $overwrite_flag = true ) {
 	protected function setup_template_vars() {
 		if ( empty( $this->repository_args ) ) {
 			$this->repository_args = $this->filter_repository_args( $this->setup_repository_args() );
+			$this->repository = tribe_events();
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1358]: :warning: Equals sign not aligned with surrounding assignments; expected 6 spaces but found 1 space

<sub>hash: dec4ee89defabdb9145aea858c671db5</sub>

File: src/Tribe/Views/V2/View.php [hunk 2]
```diff
@@ -1426,6 +1430,7 @@ protected function setup_template_vars() {
 			'breakpoint_pointer'     => $this->get_breakpoint_pointer(),
 			'is_initial_load'        => $this->context->doing_php_initial_state(),
 			'public_views'           => $this->get_public_views( $url_event_date ),
+			'show_recent_past'       => $this->should_show_recent_past_events_view(),
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:1433]: :warning: Array double arrow not aligned correctly; expected 5 space(s) between "'show_recent_past'" and double arrow, but found 7.

<sub>hash: cdf7e728c2cc672c2307533e86b317e9</sub>

File: src/Tribe/Views/V2/View.php [hunk 3]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
 
 		return $data;
 	}
+
+	protected function should_show_recent_past_events_view() {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2158]: :no_entry_sign: Missing doc comment for function should_show_recent_past_events_view()

<sub>hash: 4099bfca1befd9c496d5f51ac494a896</sub>

File: src/Tribe/Views/V2/View.php [hunk 4]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+		 * @param View    $instance The current View object.
+		 */
+		$show = apply_filters( "tribe_events_views_v2_{$this->get_slug()}_show_recent_past_events_view", $show, $this );
+
+		return $show;
+	}
+
+
+	/**
+	 * @param array $events
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2187]: :no_entry_sign: Missing parameter comment

<sub>hash: a17b00e440c0bf97be4cb7c33c922c2d</sub>

File: src/Tribe/Views/V2/View.php [hunk 5]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+	 * @param array $template_vars
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2188]: :no_entry_sign: Missing parameter comment

<sub>hash: 22dd6f87b2738de3ffb10f993f6bdf97</sub>

File: src/Tribe/Views/V2/View.php [hunk 6]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+	 */
+	protected function setup_additional_views( array $events = [], array $template_vars = [] ) {
+		// todo: flatten the events if required (month, week)
+		// todo add threshold and filter
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2192]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 685e6ce1f24cf120bf3bc60d1ce7bc43</sub>

File: src/Tribe/Views/V2/View.php [hunk 7]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+		if ( 0 === count( $events ) && ! empty( $template_vars['show_recent_past'] ) ) {
+			$template_vars['show_recent_past'] = true;
+			$recent_past_view                  = new Recent_Past_View();
+			$recent_past_view->set_context( $this->context );
+			$instance = $recent_past_view->make( Recent_Past_View::Class );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2197]: :warning: Equals sign not aligned with surrounding assignments; expected 31 spaces but found 1 space

<sub>hash: 1070b5b5acc74c3ab2860151146e7baf</sub>

File: src/Tribe/Views/V2/View.php [hunk 8]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+			$recent_past_view->template = new Template( $instance );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2198]: :warning: Equals sign not aligned with surrounding assignments; expected 13 spaces but found 1 space

<sub>hash: f13104e8b5bff96f7eb72717f4166c86</sub>

File: src/Tribe/Views/V2/View.php [hunk 9]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+			$template_vars['recent_past_view_data'] = $recent_past_view->get_template_vars();
+
+			$whitelist                              = [
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2201]: :warning: Equals sign not aligned correctly; expected 1 space but found 30 spaces

<sub>hash: a771dd69b730ccb417247d281329bb12</sub>

File: src/Tribe/Views/V2/View.php [hunk 10]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+				'components/events-bar/filters',
+				'components/events-bar/views',
+				'components/events-bar/views/list/item',
+				'list/top-bar',
+				'list/top-bar/nav',
+				'list/top-bar/datepicker',
+				'list',
+				// todo: add here the templates we need and use in the recent past events view.
+			];
+			add_filter( 'tribe_template_done', static function ( $done = null, $name ) use ( $whitelist ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2227]: * :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Arguments with default values must be at the end of the argument list

<sub>hash: baba422687b72be59cf49611aa569977</sub>

File: src/Tribe/Views/V2/View.php [hunk 11]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+				if ( in_array( $name, $whitelist, true ) ) {
+					return $done;
+				}
+
+				return '';
+			}, 1, 2 );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2233]: * :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: ec0bd46731432fd4c51c3d5ed55811a6</sub>

File: src/Tribe/Views/V2/View.php [hunk 12]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+
+			//todo hook into "tribe_template_html:components/messages" start filtering templates
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2235]: :no_entry_sign: No space found before comment text; expected "// todo hook into "tribe_template_html:components/messages" start filtering templates" but found "//todo hook into "tribe_template_html:components/messages" start filtering templates"

<sub>hash: cad17b03526cb76f2aca81842bf92138</sub>

File: src/Tribe/Views/V2/View.php [hunk 13]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+			//apply_filters( "tribe_template_html:$hook_name", $html, $file, $name, $this )
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2236]: * :no_entry_sign: No space found before comment text; expected "// apply_filters( "tribe_template_html:$hook_name", $html, $file, $name, $this )" but found "//apply_filters( "tribe_template_html:$hook_name", $html, $file, $name, $this )"
* :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: a5b8abc813c1dbbf3a15f96aac12b35e</sub>

File: src/Tribe/Views/V2/View.php [hunk 14]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+
+			//todo hook into "tribe_template_html:month" - render
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2238]: * :no_entry_sign: No space found before comment text; expected "// todo hook into "tribe_template_html:month" - render" but found "//todo hook into "tribe_template_html:month" - render"
* :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: cb5350e30189224d6ad8b2921a7fe5f2</sub>

File: src/Tribe/Views/V2/View.php [hunk 15]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+			add_filter( 'tribe_template_html:events/v2/list', static function ( $html ) use ( $recent_past_view ) {
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2239]: * :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: 76dd987a0f707a72e7a5cf0c80c21ecb</sub>

File: src/Tribe/Views/V2/View.php [hunk 16]
```diff
@@ -2145,4 +2154,92 @@ protected function get_container_data() {
...
+
+				return $html . $recent_past_view->get_html();
+			} );
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/View.php:2242]: :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: 0e419aae1298fb2d11da930744da449a</sub>

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+use Tribe\Events\Views\V2\View;
+use Tribe\Events\Views\V2\Views\Traits\List_Behavior;
+use Tribe__Context;
+use Tribe__Events__Main as TEC;
+use Tribe__Events__Rewrite as TEC_Rewrite;
+use Tribe__Utils__Array as Arr;
+use Tribe\Events\Views\V2\Utils;
+
+class Recent_Past_View extends View {
+	use List_Behavior;
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:20]: :no_entry_sign: Missing doc comment for class Recent_Past_View

<sub>hash: bf49437c46abd745ddaaee83a0856361</sub>

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+	 * Indicates Recent Past View supports the date as a query argument appended to its URL, not as part of a "pretty" URL.
+	 *
+	 * @var bool
+	 */
+	protected static $date_in_url = false;
+
+		// Todo: remove this.
+
+	public function maybe_filter_current_view_slug( $view_slug, $context, $query ) {
+
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:49]: :no_entry_sign: You must use "/**" style comments for a function comment

<sub>hash: 7dffdf6922b77a58ede29385330888a9</sub>

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+		if ( tribe_events()->where( 'ends_after', 'now' )->count() === 0 ) {
+
+			return 'recent-past';
+		}
+
+		return $view_slug;
+	}
+
+	/**
+	 * {@inheritDoc}
```
Comment (tr1b0t) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:72]: :no_entry_sign: Doc comment for parameter "$context" missing

<sub>hash: 7cec6880dfb0481a7b65a42e6d546e6b</sub>

File: src/views/v2/recent-past.php [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+?>
+<div
+	<?php tribe_classes( $container_classes ); ?>
+	data-js="tribe-events-view"
+	data-view-rest-nonce="<?php echo esc_attr( $rest_nonce ); ?>"
+	data-view-rest-url="<?php echo esc_url( $rest_url ); ?>"
+	data-view-manage-url="<?php echo esc_attr( $should_manage_url ); ?>"
+	<?php foreach ( $container_data as $key => $value ) : ?>
```
Comment (tr1b0t) [inline src/views/v2/recent-past.php:31]: :no_entry_sign: Wrong escaping function. href and src attributes should be escaped by `esc_url()`, not by `esc_attr()`.

<sub>hash: b707258a560c7836d2fae404ff3f9208</sub>

File: src/views/v2/recent-past.php [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+		data-view-<?php echo esc_attr( $key ) ?>="<?php echo esc_attr( $value ) ?>"
+	<?php endforeach; ?>
```
Comment (tr1b0t) [inline src/views/v2/recent-past.php:33]: * :no_entry_sign: Inline PHP statement must end with a semicolon
* :no_entry_sign: Inline PHP statement must end with a semicolon

<sub>hash: f68f544d82fda78cd17cc7a376c8f3a4</sub>

File: src/views/v2/recent-past.php [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+
+				<?php $this->template( 'list/event', [ 'event' => $event ] ); ?>
+
+			<?php endforeach; ?>
+
+		</div>
+
+	</div>
+</div>
\ No newline at end of file
```
Comment (tr1b0t) [inline src/views/v2/recent-past.php:53]: :no_entry_sign: File must end with a newline character

<sub>hash: d3e572a7169772f645d0461e49193553</sub>

File: src/Tribe/Views/V2/View.php [hunk 23]
```diff
@@ -2145,4 +2149,58 @@ protected function get_container_data() {
...
+
+	/**
+	 * Filters Whether the Recent Past Events Should Show for a specific View.
+	 *
+	 * @since TBD
+	 *
+	 * @return mixed|void
+	 */
+	protected function should_show_recent_past_events_view() {
+		$show = $this->context->get( 'show_recent_past', true );
```
Comment (lucatume) [inline src/Tribe/Views/V2/View.php:2161]: I just want to add a note for anyone reviewing this that might look for a `show_recent_past` location added to the Context: there is none and it's fine.

This will allow us, in the future, to add an option and expose a UI to toggle if the Recent Past Events view should show or not and add the corresponding Context location then.

File: src/Tribe/Views/V2/View.php [hunk 24]
```diff
@@ -2145,4 +2149,58 @@ protected function get_container_data() {
...
+	 * Setup of Additional Views into another View.
+	 *
+	 * @since TBD
+	 *
+	 * @param array $events        Array that will be counted to verify if we have events.
+	 * @param array $template_vars An associative array of variables that will be set, and exported, in the template.
+	 */
+	protected function setup_additional_views( array $events = [], array $template_vars = [] ) {
+
+		// todo: flatten the events if required (month, week)
```
Comment (lucatume) [inline src/Tribe/Views/V2/View.php:2197]: So this one still has to be addressed, right @jesseeproductions?

File: src/Tribe/Views/V2/View.php [hunk 25]
```diff
@@ -2145,4 +2149,58 @@ protected function get_container_data() {
...
+		// todo add threshold and filter
```
Comment (lucatume) [inline src/Tribe/Views/V2/View.php:2198]: This one should be removed as it's addressed in the code below it.

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+	/**
+	 * Visibility for this view.
+	 *
+	 * @since TBD
+	 *
+	 * @var bool
+	 */
+	protected static $publicly_visible = true;
+
+	protected $whitelist = [
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:36]: Missing doc block here.

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+		'components/top-bar/actions',
+		'components/events-bar/search/keyword',
+		'components/events-bar/search/submit',
+		'components/events-bar/search-button',
+		'components/events-bar/tabs',
+		'components/events-bar/search',
+		'components/events-bar/filters',
+		'components/events-bar/views',
+
+		// List View.
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:56]: Complete this for the rest of the Views.

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+
+		return $args;
+	}
+
+	/**
+	 * Add Filters for Whitelist and Adding View HTML.
+	 *
+	 * @since TBD
+	 */
+	public function hook_whitelist_and_html_template() {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:104]: I would rename this to `add_view_filters`.
While we only have this view as an additional View, we might very well have more in the future.
Then we'll refactor additional views to extend an `Addtional_View_Abstract` class of sorts, then a more generic name will allow us to group a lot of common logic in that abstract class and leave the Additional Views implementations deal only with the "configuration".

TL;DR: rename to `add_view_filters`.

File: src/Tribe/Views/V2/Views/Recent_Past_View.php [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+	 * @since TBD
+	 *
+	 * @param string  $done    Whether to continue displaying the template or not.
+	 * @param array   $name    Template name.
+	 * @param array   $context Any context data you need to expose to this file.
+	 * @param boolean $echo    If we should also print the Template.
+	 *
+	 * @return string
+	 */
+	public function filter_template_display_by_whitelist( $done, $name, $context, $echo ) {
```
Comment (lucatume) [inline src/Tribe/Views/V2/Views/Recent_Past_View.php:132]: Rename this to `filter_template_done`?

**Comment by lucatume** [comment]
@jesseeproductions Should this be put in draft?

File: src/Tribe/Views/V2/Views/Recent_Past_View.php (hunk 28 above)
Comment (jesseeproductions) [reply src/Tribe/Views/V2/Views/Recent_Past_View.php:104]: 👍

**Comment by jesseeproductions** [comment]
@lucatume can this be set to draft? I did not see a way to go back.
Overall it is working in every view. 
Except there is an issue in Month view to work out. In April it shows three March days. If those days have events it uses those in the total count. So I have to add to the coding a way to only get the current month's events to prevent that.

**Comment by lucatume** [comment]
> @lucatume can this be set to draft? I did not see a way to go back.

Me neither, "hold" status is fine.


> Except there is an issue in Month view to work out. In April it shows three March days. If those days have events it uses those in the total count. So I have to add to the coding a way to only get the current month's events to prevent that.

You will probably need to do the flattening of events for nested events-per-day as I've done to setup messages in the `View` class.

File: tests/views_integration/Tribe/Events/Views/V2/Views/__snapshots__/Day_ViewTest__test_render_empty__1.php [hunk 30]
```diff
@@ -156,6 +156,17 @@ class="tribe-events-c-view-selector__list-item-link"
...
+</li>
+					<li class="tribe-events-c-view-selector__list-item tribe-events-c-view-selector__list-item--latest-past">
+	<a
+		href="http://test.tri.be/?post_type=tribe_events&#038;eventDisplay=latest-past&#038;tribe-bar-date=2019-01-01"
+		class="tribe-events-c-view-selector__list-item-link"
+		data-js="tribe-events-view-link"
+	>
+		<span class="tribe-events-c-view-selector__list-item-icon tribe-common-svgicon tribe-common-svgicon--latest-past"></span>
+		<span class="tribe-events-c-view-selector__list-item-text">
+			Latest-past		</span>
```
Comment (paulmskim) [inline tests/views_integration/Tribe/Events/Views/V2/Views/__snapshots__/Day_ViewTest__test_render_empty__1.php:168]: This should not have a dash on it. Where are we setting this, can we make sure that it says Latest Past instead?
